### DISCUSSION TITLE
Support for ECMAScript Decorators?

### DISCUSSION THREAD

**Discussion by hasezoey** [Automattic/mongoose#13872] [General]
Since [Decorators (`@decorator`)](https://github.com/tc39/proposal-decorators) are almost standardized (stage 3), should mongoose consider supporting them, where maybe the function functionality of `.loadClass` is extended? (maybe even more similar to how `typegoose` currently works?)

<details>
<summary>Example of decorator usage from typegoose</summary>

```ts
class Job {
  @prop()
  public title?: string;

  @prop()
  public position?: string;
}

class Car {
  @prop()
  public model?: string;
}

class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; // This is a single Primitive

  @prop({ type: () => [String] })
  public preferences?: string[]; // This is a Primitive Array

  @prop()
  public mainJob?: Job; // This is a single SubDocument

  @prop({ type: () => [Job] })
  public jobs?: Job[]; // This is a SubDocument Array

  @prop({ ref: () => Car })
  public mainCar?: Ref<Car>; // This is a single Reference

  @prop({ ref: () => [Car] })
  public cars?: Ref<Car>[]; // This is a Reference Array
}
```

</details>

Note: this thread is made to get to know if mongoose wants to make use of decorators first-party (instead of extensions like `typegoose`)

**Comment by vkarpov15** [comment]
When decorators are formally included in the ECMAScript spec and fully supported in Node.js, we will consider using decorators in Mongoose. Stage 3 proposals have fallen through before, so I don't want to put the effort in to support decorators until they're formally adopted.
